pkg/astrolabe/data_transport.go

```diff
@@ -44,27 +52,113 @@ func NewDataTransport(transportType string, params map[string]string) DataTransp
 	}
 }
 
+const(
+	S3TransportType = "s3"
+	S3URLParam = "url"
+	S3HostParam = "host"
+	S3BucketParam = "bucket"
+	S3KeyParam = "key"
+)
+
+type S3Config struct {
+	Port int			`json:"port,omitempty"`
+	Host net.IP			`json:"host,omitempty"`
+	AccessKey string	`json:"accessKey,omitempty"`
+	Secret string		`json:"secret,omitempty"`
+	Prefix string		`json:"prefix,omitempty"`
+	URLBase string		`json:"urlBase,omitempty"`
+}
+
+func getLocalIP() (net.IP, error) {
+	addrs, err := net.InterfaceAddrs()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, a := range addrs {
+		if ipnet, ok := a.(*net.IPNet); ok && !ipnet.IP.IsLoopback() {
+			if ipnet.IP.To4() != nil {
+				return ipnet.IP, nil
+			}
+		}
+	}
+	return nil, errors.New("Could not determine local IP address")
+}
+
+func (this S3Config) getURL() (string) {
+	return "http://" + this.Host.String() + ":" + strconv.Itoa(this.Port) + "/"+this.Prefix+"/"
+}
+
 func NewDataTransportForS3URL(url string) DataTransport {
 	return DataTransport{
-		transportType: "s3",
+		transportType: S3TransportType,
 		params: map[string]string{
-			"url": url,
+			S3URLParam: url,
 		},
 	}
 }
 
 func NewDataTransportForS3(host string, bucket string, key string) DataTransport {
 	url := "http://" + host + "/" + bucket + "/" + key
 	return DataTransport{
-		transportType: "s3",
+		transportType: S3TransportType,
 		params: map[string]string{
-			"url":    url,
-			"host":   host,
-			"bucket": bucket,
-			"key":    key,
+			S3URLParam:    url,
+			S3HostParam:   host,
+			S3BucketParam: bucket,
+			S3KeyParam:    key,
 		},
 	}
 }
+
+const (
+	DataExt = ""
+	MDExt = ".md"
+	CombinedExt = ".zip"
+	PEInfoExt = ".peinfo"
+)
+
+func NewS3DataTransportForPEID(peid ProtectedEntityID, s3Config S3Config) DataTransport {
+	return NewS3TransportForPEID(peid, DataExt, s3Config)
+}
+
+func NewS3MDTransportForPEID(peid ProtectedEntityID, s3Config S3Config) DataTransport {
+	return NewS3TransportForPEID(peid, MDExt, s3Config)
+}
+
+func NewS3CombinedTransportForPEID(peid ProtectedEntityID, s3Config S3Config) DataTransport {
+	return NewS3TransportForPEID(peid, CombinedExt, s3Config)
+}
+
+func NewS3PEInfoTransportForPEID(peid ProtectedEntityID, s3Config S3Config) DataTransport {
+	return NewS3TransportForPEID(peid, PEInfoExt, s3Config)
+}
+
+
+func NewS3TransportForPEID(peid ProtectedEntityID, ext string, s3Config S3Config) DataTransport {
+	credentials := credentials2.NewStaticCredentials(s3Config.AccessKey, s3Config.Secret, "")
+	s3ForcePathStyle := true
+	sess, err := session.NewSession(&aws.Config{
+		Endpoint: aws.String(s3Config.getURL()),
+		Credentials: credentials,
+		Region: aws.String("us-west-2"),
+		S3ForcePathStyle: &s3ForcePathStyle,
+	},
+	)
+
+	if err != nil {
+
```

Comment:
missing return?

Reply:
Yah, you're right.  I need to plumb this for errors the whole way.  I miss exceptions